libraries/AP_HAL_PX4/CAN.cpp

```diff
@@ -0,0 +1,1035 @@
+{
+    uint64_t usec = 0;
+    usec = AP_HAL::micros64();
+
+    return uavcan::MonotonicTime::fromUSec(usec);
+}
+
+BusEvent::BusEvent(PX4CANDriver& can_driver) :
+    _signal(0)
+{
+}
+
+BusEvent::~BusEvent()
+{
+}
+
+bool BusEvent::wait(uavcan::MonotonicDuration duration)
+{
+    struct hrt_call wait_call;
+
+    irqstate_t irs = irqsave();
+    if (_signal) {
+        _signal = 0;
+        irqrestore(irs);
+        return true;
+    }
+
+    sem_init(&_wait_semaphore, 0, 0);
+    irqrestore(irs);
+
+    hrt_call_after(&wait_call, duration.toUSec(), (hrt_callout) signalFromCallOut, this);
+    sem_wait(&_wait_semaphore);
+
+    hrt_cancel(&wait_call);
+
+    irs = irqsave();
+    if (_signal) {
+        _signal = 0;
+        irqrestore(irs);
+
+        return true;
+    }
+    irqrestore(irs);
+
+    return false;
+}
+
+void BusEvent::signalFromCallOut(BusEvent *sem)
+{
+    sem_post(&sem->_wait_semaphore);
+}
+
+void BusEvent::signalFromInterrupt()
+{
+    _signal++;
+    sem_post(&_wait_semaphore);
+}
+
+static void handleTxInterrupt(uint8_t iface_index)
+{
+    if (iface_index < UAVCAN_STM32_NUM_IFACES) {
+        uint64_t utc_usec = clock::getUtcUSecFromCanInterrupt();
+        if (utc_usec > 0) {
+            utc_usec--;
+        }
+        PX4CAN* iface = ((PX4CANDriver*) hal.can_mgr)->getIface(iface_index);
+        if (iface != UAVCAN_NULLPTR) {
+            iface->handleTxInterrupt(utc_usec);
+        }
+    }
+}
+
+static void handleRxInterrupt(uint8_t iface_index, uint8_t fifo_index)
+{
+    if (iface_index < UAVCAN_STM32_NUM_IFACES) {
+        uint64_t utc_usec = clock::getUtcUSecFromCanInterrupt();
+        if (utc_usec > 0) {
+            utc_usec--;
+        }
+        PX4CAN* iface = ((PX4CANDriver*) hal.can_mgr)->getIface(iface_index);
+        if (iface != UAVCAN_NULLPTR) {
+            iface->handleRxInterrupt(fifo_index, utc_usec);
+        }
+    }
+}
+
+const uint32_t PX4CAN::TSR_ABRQx[PX4CAN::NumTxMailboxes] = { bxcan::TSR_ABRQ0, bxcan::TSR_ABRQ1, bxcan::TSR_ABRQ2 };
+PX4CAN* PX4CANDriver::ifaces[UAVCAN_STM32_NUM_IFACES] = { UAVCAN_NULLPTR, UAVCAN_NULLPTR };
+
+int PX4CAN::computeTimings(const uint32_t target_bitrate, Timings& out_timings)
+{
+    if (target_bitrate < 1) {
+        return -ErrInvalidBitRate;
+    }
+
+    /*
+     * Hardware configuration
+     */
+    const uint32_t pclk = STM32_PCLK1_FREQUENCY;
+
+    static const int MaxBS1 = 16;
+    static const int MaxBS2 = 8;
+
+    /*
+     * Ref. "Automatic Baudrate Detection in CANopen Networks", U. Koppe, MicroControl GmbH & Co. KG
+     *      CAN in Automation, 2003
+     *
+     * According to the source, optimal quanta per bit are:
+     *   Bitrate        Optimal Maximum
+     *   1000 kbps      8       10
+     *   500  kbps      16      17
+     *   250  kbps      16      17
+     *   125  kbps      16      17
+     */
+    const int max_quanta_per_bit = (target_bitrate >= 1000000) ? 10 : 17;
+
+    if (!(max_quanta_per_bit <= (MaxBS1 + MaxBS2))) {
```

Comment:
I don't understand. The check in code is equivalent to (and this is what I propose that should be there):

`if (max_quanta_per_bit > (MaxBS1 + MaxBS2))`

So where does the > 999999 come from?